Fix setting sizes attribute on PICTURE > SOURCE elements

mukeshpanchal27 opened the pull request:
## Summary

<!-- Please reference the issue this PR fixes. If this PR does not fix the entire issue, change this to Addresses #... instead. -->
Fixes #1349

The PR changes the filter priority for the picture element filter so it will be generated before the sizes attribute updates. Add new filter for picture element in auto-sizes plugin that add accurate sizes change for `source ` tags

<!--
For maintainers only, please make sure:

- PR has a `[Type]` label.
- PR has a plugin-specific milestone, or the `no milestone` label if it does not apply to any specific plugin.
- PR has a changelog-friendly title, or the `skip changelog` label if it should not be mentioned in the plugin's changelog.
-->

mukeshpanchal27 committed:
Fix sizes issue for picture element

mukeshpanchal27 committed:
Merge branch 'trunk' into fix/1349-sizes-picture-element

westonruter commented:
Can a test be added which demonstrates the bug if the changes aren't applied?

westonruter commented on plugins/auto-sizes/hooks.php:
```diff
@@ -216,3 +216,37 @@ function auto_sizes_improve_image_sizes_attributes( string $content ): string {
+		$sizes = $processor->get_attribute( 'sizes' );
+	}
+
+	if ( ! $sizes ) {
```
In anticipation of strict rules: #1241 

```suggestion
	if ( ! is_string( $sizes ) ) {
```

westonruter commented on plugins/auto-sizes/hooks.php:
```diff
@@ -216,3 +216,37 @@ function auto_sizes_improve_image_sizes_attributes( string $content ): string {
+		return $content;
+	}
+
+	$has_image = $processor->next_tag( array( 'tag_name' => 'img' ) );
+	$sizes     = null;
+	// Only update the markup if an image is found.
+	if ( $has_image ) {
+		$sizes = $processor->get_attribute( 'sizes' );
+	}
```
This can be simplified:

```suggestion
	// Only update the markup if an image is found.
	if ( ! $processor->next_tag( array( 'tag_name' => 'img' ) ) ) {
		return $content;
	}
	$sizes = $processor->get_attribute( 'sizes' );
```

westonruter commented on plugins/auto-sizes/hooks.php:
```diff
@@ -216,3 +216,37 @@ function auto_sizes_improve_image_sizes_attributes( string $content ): string {
+
+	$processor = new WP_HTML_Tag_Processor( $content );
+	while ( $processor->next_tag( array( 'tag_name' => 'source' ) ) ) {
+		$processor->set_attribute( 'sizes', $sizes );
```
What if there is already a `sizes` attribute on this `source` tag? Shouldn't it do nothing in that case?

```suggestion
		if ( ! is_string( $processor->get_attribute( 'sizes' ) ) ) {
			$processor->set_attribute( 'sizes', $sizes );
		}
```

mukeshpanchal27 replied:
The `sizes` attribute is already set https://github.com/WordPress/performance/blob/trunk/plugins/webp-uploads/picture-element.php#L130 for `source` tags. We are updating it to be accurate based on the new API changes.

We take the correct `sizes` value from the `img` tag and update it in the `source` tags.

westonruter commented on plugins/auto-sizes/hooks.php:
```diff
@@ -216,3 +216,37 @@ function auto_sizes_improve_image_sizes_attributes( string $content ): string {
+	}
+
+	$processor = new WP_HTML_Tag_Processor( $content );
+	while ( $processor->next_tag( array( 'tag_name' => 'source' ) ) ) {
```
Question: Is the `srcset` getting set on each of the `source` tags, but the `sizes` attribute is not? Why?

mukeshpanchal27 replied:
Both `srcset` and `sizes` set through picture element code but we are updating only `sizes` to more accurate based on our new API.

mukeshpanchal27 committed:
Merge branch 'trunk' into fix/1349-sizes-picture-element

mukeshpanchal27 committed:
Use strict tules

Co-authored-by: Weston Ruter <westonruter@google.com>

mukeshpanchal27 committed:
Simplified code

Co-authored-by: Weston Ruter <westonruter@google.com>

mukeshpanchal27 commented:
> Can a test be added which demonstrates the bug if the changes aren't applied?

@westonruter the bug only generated when the Enhanced Responsive Images (auto-sizes) plugin also activated.

joemcgill reviewed (changes requested):
I don't think this is the most robust approach to fixing this issue. If I am understanding the problem correctly, the issue is caused because the Enhanced Responsive Images plugin modifies the `sizes` attribute using the `wp_content_img_tag` filter, since we're not getting all of the information we need to filter the `wp_calculate_image_sizes()` value directly. 

However, when Modern Image Formats applies `webp_uploads_wrap_image_in_picture()` it adds `sizes` attributes to the `source` elements by calling `wp_calculate_image_sizes()` directly, rather than using the actual `sizes` attribute that is on the `img` element that it is wrapping. 

I think that a better solution would be to update `webp_uploads_wrap_image_in_picture()` to try to copy the `sizes` attribute from the `img` that it's wrapping, if available, rather than using `wp_calculate_image_sizes()` directly. This would ensure that any modifications that were made to the `sizes` attribute, regardless of what filter was used, will be applied to the `source` elements as well.

Alternately, we need to make sure the enhanced `sizes` attribute ends up getting applied via the `wp_calculate_image_sizes` filter so that any other code that calls `wp_calculate_image_sizes()` directly will have the optimizations applied as well.

As a side note: when encountering compatibility issues like this one, we should be skeptical about modifying filter order as a way of fixing the problem, since we can't control the filter order of all third party code that might modify the HTML markup that our plugins affect.

mukeshpanchal27 committed:
Revert auto-sizes plugin changes

mukeshpanchal27 committed:
Cleanup

mukeshpanchal27 commented:
Thanks @joemcgill for the feedback.

I have updated PR that use `sizes` value from the IMG tag and reverted the filter priority changes. I will open issue for `wp_calculate_image_sizes()`

mukeshpanchal27 committed:
Merge branch 'trunk' into fix/1349-sizes-picture-element

mukeshpanchal27 committed:
Minor docblock update

westonruter commented on plugins/webp-uploads/picture-element.php:
```diff
@@ -95,12 +95,10 @@ function webp_uploads_wrap_image_in_picture( string $image, string $context, int
-			continue;
-		}
+	// Get the sizes from the IMG tag.
+	$sizes = (string) $processor->get_attribute( 'sizes' );
```
Note that `$processor->get_attribute( 'sizes' )` here could return `null` if no attribute is present (or `true` in the case of `<img sizes>`, which is unlikely). Therefore, I recommend removing the `(string)` cast and instead when using the `$sizes` attribute to only add it to the `$picture_sources` markup if `is_string( $sizes )`.

```suggestion
	$sizes = $processor->get_attribute( 'sizes' );
```

westonruter replied:
```suggestion
	$sizes = $processor->get_attribute( 'sizes' );
```

westonruter commented on plugins/webp-uploads/tests/test-picture-element.php:
```diff
@@ -125,4 +125,75 @@ public function data_provider_it_should_maybe_wrap_images_in_picture_element():
+		// Picture filter "webp_uploads_wrap_image_in_picture" is run prior to "wp_content_img_tag".
+		$the_image = apply_filters( 'wp_content_img_tag', $the_image, 'the_content', $attachment_id );
+
+		// 1024px sizes comes from "source" tags.
+		$this->assertStringContainsString( 'sizes="(max-width: 1024px) 100vw, 1024px"', $the_image );
+		// 333px sizes comes from "IMG" tag.
+		$this->assertStringContainsString( 'sizes="(max-width: 333px) 100vw, 333px"', $the_image );
```
Let's have this do an `assertSame()` assertion to test the entire string. See https://github.com/WordPress/performance/pull/1328#discussion_r1667070230

mukeshpanchal27 replied:
Updated.

westonruter commented on plugins/webp-uploads/tests/test-picture-element.php:
```diff
@@ -125,4 +125,75 @@ public function data_provider_it_should_maybe_wrap_images_in_picture_element():
+
+		$the_image = apply_filters( 'wp_content_img_tag', $the_image, 'the_content', $attachment_id );
+
+		// 1024px sizes not set in "source" tags.
+		$this->assertStringContainsString( 'sizes="(max-width: 1024px) 100vw, 1024px"', $the_image );
+		// 333px sizes set in "IMG" and "source" tag.
+		$this->assertStringContainsString( 'sizes="(max-width: 333px) 100vw, 333px"', $the_image );
```
I don't understand. These assertions are the same as the two assertions above before adding the additional `add_filter()` call.

mukeshpanchal27 replied:
Moved to second test

westonruter commented on plugins/webp-uploads/tests/test-picture-element.php:
```diff
@@ -125,4 +125,75 @@ public function data_provider_it_should_maybe_wrap_images_in_picture_element():
+				if ( ! $processor->next_tag( array( 'tag_name' => 'img' ) ) ) {
+					return $content;
+				}
+				$processor->set_attribute( 'sizes', '(max-width: 333px) 100vw, 333px' );
```
This closure is identical to the above one. It should be saved in a variable and then re-used in both `add_filter()` calls then.

westonruter commented on plugins/webp-uploads/tests/test-picture-element.php:
```diff
@@ -125,4 +125,75 @@ public function data_provider_it_should_maybe_wrap_images_in_picture_element():
 		);
 	}
+
+	public function test_sizes_value_for_picture_element(): void {
```
Please also add a test for when an `img` doesn't have `sizes` at all.

westonruter commented on plugins/webp-uploads/tests/test-picture-element.php:
```diff
@@ -125,4 +125,75 @@ public function data_provider_it_should_maybe_wrap_images_in_picture_element():
+		// 333px sizes comes from "IMG" tag.
+		$this->assertStringContainsString( 'sizes="(max-width: 333px) 100vw, 333px"', $the_image );
+
+		// Run filter prior to Picture filter "webp_uploads_wrap_image_in_picture".
```
Should this perhaps be moved to another test? It seems like this test could be divided in two.

mukeshpanchal27 committed:
Remove casting

Co-authored-by: Weston Ruter <westonruter@google.com>

mukeshpanchal27 committed:
Fix phpstan

mukeshpanchal27 committed:
Merge branch 'trunk' into fix/1349-sizes-picture-element

mukeshpanchal27 committed:
Address unit tests feedback

mukeshpanchal27 committed:
Add tests without sizes and srcset

mukeshpanchal27 commented:
Thanks @westonruter for the feedback. PR ready for review.

joemcgill commented on plugins/webp-uploads/tests/test-picture-element.php:
```diff
@@ -125,4 +125,229 @@ public function data_provider_it_should_maybe_wrap_images_in_picture_element():
+		$this->opt_in_to_picture_element();
+
+		// Create an image.
+		$attachment_id = self::factory()->attachment->create_upload_object( TESTS_PLUGIN_DIR . '/tests/data/images/leaves.jpg' );
```
We're duplicating this in several tests, which will be a bit slow. Could we move this process to a shared fixture using the `wpSetUpBeforeClass()` method and store the attachment ID as a static property? See an example [here](https://github.com/WordPress/performance/blob/fix/1349-sizes-picture-element/plugins/auto-sizes/tests/test-auto-sizes.php#L11-L23).

mukeshpanchal27 replied:
I have used the shared fixture for the new test. However, if I use it in the old test, I will have to refactor the entire test or update it. We can address this in a follow-up PR.

joemcgill commented on plugins/webp-uploads/tests/test-picture-element.php:
```diff
@@ -125,4 +125,229 @@ public function data_provider_it_should_maybe_wrap_images_in_picture_element():
+		);
+
+		$expected_html = '<picture class="wp-picture-{{img-attachment-id}}" style="display: contents;"><source type="image/webp" srcset="{{webp-srcset}}" sizes="{{source-sizes}}"><source type="image/jpeg" srcset="{{jpeg-srcset}}" sizes="{{source-sizes}}"><img width="{{img-width}}" height="{{img-height}}" src="{{img-src}}" class="wp-image-{{img-attachment-id}}" alt="{{img-alt}}" decoding="async" loading ...
+		$expected_html = str_replace( array_keys( $replacements ), array_values( $replacements ), $expected_html );
```
There are a lot of implementation details in these tests, which seems unnecessary, and makes it difficult to understand what is actually being tested.

I think all we're trying to test is whether the `source` elements generated by this plugin include the same `sizes` attribute that the original `img` element has after being filtered, it would be more straightforward to generate the image markup without picture support, then again with picture applied and assert that the `sizes` attributes of the `source` elements are the same as the `sizes` attribute of the `img` without picture support. Here's an untested example of what I would expect:

```php
$image = wp_get_attachment_image( $attachment_id );

$img_markup = apply_filters( 'the_content', $image );

// Apply picture element support.
$this->opt_in_to_picture_element();

$picture_markup = apply_filters( 'the_content', $image );

$img_processor = new WP_HTML_Tag_Processor( $img_markup );
$img_processor->next_tag( array( 'tag_name' => 'IMG' );
$img_sizes = $img_processor->get_attribute( 'sizes' );

$picture_processor = new WP_HTML_Tag_Processor( $picture_markup );
$picture_processor->next_tag( array( 'tag_name' => 'source' ) ) );
$picture_source_sizes = $picture_processor->get_attribute( 'sizes' );

$this->assertSame( $img_sizes, $picture_source_sizes );
```
Then you would need another test that changes the sizes attribute of the original markup before applying `the_content` filters to make sure the `sizes` attributes are still the same.

The same is true for the other tests in this class.

mukeshpanchal27 replied:
Updated in new test.

joemcgill commented on plugins/webp-uploads/tests/test-picture-element.php:
```diff
@@ -125,4 +125,229 @@ public function data_provider_it_should_maybe_wrap_images_in_picture_element():
+		$this->assertSame( $expected_html, $the_image );
+	}
+
+	public function test_that_img_sizes_is_equal_to_source_for_picture_elementsss(): void {
```
Not sure what this is testing, but the name could be more descriptive.

joemcgill reviewed (dismissed):
This looks like it's working, but the tests could use some improvements. I've left some suggestions but let me know if they're not clear.

adamsilverstein commented on plugins/webp-uploads/picture-element.php:
```diff
@@ -124,10 +122,10 @@ function webp_uploads_wrap_image_in_picture( string $image, string $context, int
 			esc_attr( $image_mime_type ),
-			esc_attr( $image_srcset ),
-			esc_attr( $sizes )
+			! empty( $image_srcset ) ? sprintf( ' srcset="%s"', esc_attr( $image_srcset ) ) : '',
```
Great! love this approach.

adamsilverstein commented on plugins/webp-uploads/picture-element.php:
```diff
@@ -95,12 +95,10 @@ function webp_uploads_wrap_image_in_picture( string $image, string $context, int
-			continue;
-		}
+	// Get the sizes from the IMG tag.
+	$sizes = $processor->get_attribute( 'sizes' );
```
Nice!

westonruter commented on plugins/webp-uploads/picture-element.php:
```diff
@@ -124,10 +122,10 @@ function webp_uploads_wrap_image_in_picture( string $image, string $context, int
 			esc_attr( $image_mime_type ),
-			esc_attr( $image_srcset ),
-			esc_attr( $sizes )
+			! empty( $image_srcset ) ? sprintf( ' srcset="%s"', esc_attr( $image_srcset ) ) : '',
```
For PHPStan strict rules compliance:

```suggestion
			is_string( $image_srcset ) ? sprintf( ' srcset="%s"', esc_attr( $image_srcset ) ) : '',
```

mukeshpanchal27 replied:
Updated in https://github.com/WordPress/performance/pull/1354/commits/6954f44727b86e1a25acee32432982b75c76469c

mukeshpanchal27 committed:
Merge branch 'trunk' into fix/1349-sizes-picture-element

mukeshpanchal27 committed:
Update unit tests

mukeshpanchal27 committed:
PHPStan strict rules

mukeshpanchal27 committed:
Use shared fixtures for attachment

westonruter commented on plugins/webp-uploads/tests/test-picture-element.php:
```diff
@@ -125,4 +140,124 @@ public function data_provider_it_should_maybe_wrap_images_in_picture_element():
+		$picture_markup = apply_filters( 'the_content', $image );
+
+		$img_processor = new WP_HTML_Tag_Processor( $img_markup );
+		$img_processor->next_tag( array( 'tag_name' => 'IMG' ) );
```
```suggestion
		$this->assertTrue( $img_processor->next_tag( array( 'tag_name' => 'IMG' ) ) );
```

westonruter commented on plugins/webp-uploads/tests/test-picture-element.php:
```diff
@@ -125,4 +140,124 @@ public function data_provider_it_should_maybe_wrap_images_in_picture_element():
+		$img_sizes = $img_processor->get_attribute( 'sizes' );
+
+		$picture_processor = new WP_HTML_Tag_Processor( $picture_markup );
+		$picture_processor->next_tag( array( 'tag_name' => 'source' ) );
+		$picture_source_sizes = $picture_processor->get_attribute( 'sizes' );
+
+		$this->assertSame( $img_sizes, $picture_source_sizes );
```
There will be more than one `source` tag, right? As such perhaps better to use a `while()` loop, and finally to check that the `img` at the end of the `picture` also has the same expected `sizes`:

```suggestion
		while ( $picture_processor->next_tag( array( 'tag_name' => 'source' ) ) ) {
			$this->assertSame( $img_sizes, $picture_processor->get_attribute( 'sizes' ) );
		}
		$this->assertTrue( $img_processor->next_tag( array( 'tag_name' => 'IMG' ) ) );
		$this->assertSame( $img_sizes, $picture_processor->get_attribute( 'sizes' ) 
```

westonruter commented on plugins/webp-uploads/tests/test-picture-element.php:
```diff
@@ -125,4 +140,124 @@ public function data_provider_it_should_maybe_wrap_images_in_picture_element():
+
+		$picture_markup = apply_filters( 'the_content', $image );
+
+		$img_processor = new WP_HTML_Tag_Processor( $img_markup );
+		$img_processor->next_tag( array( 'tag_name' => 'IMG' ) );
+		$img_sizes = $img_processor->get_attribute( 'sizes' );
+
+		$picture_processor = new WP_HTML_Tag_Processor( $picture_markup );
+		$picture_processor->next_tag( array( 'tag_name' => 'source' ) );
+		$picture_source_sizes = $picture_processor->get_attribute( 'sizes' );
+
+		$this->assertSame( $img_sizes, $picture_source_sizes );
```
See suggestions for the same code below.

westonruter commented on plugins/webp-uploads/tests/test-picture-element.php:
```diff
@@ -125,4 +140,124 @@ public function data_provider_it_should_maybe_wrap_images_in_picture_element():
+			)
+		);
+
+		// Run filter prior to Picture filter "webp_uploads_wrap_image_in_picture".
+		add_filter(
+			'wp_content_img_tag',
+			static function (
+				string $content
+			): string {
+				$processor = new WP_HTML_Tag_Processor( $content );
+				if ( ! $processor->next_tag( array( 'tag_name' => 'img' ) ) ) {
+					return $content;
+				}
+				$processor->set_attribute( 'sizes', '(max-width: 333px) 100vw, 333px' );
+				return $processor->get_updated_html();
+			}
+		);
```
This is the only change to this test compared with the previous test, correct? As such, maybe better to use a data provider and have this `add_filter()` call be inside of a closure that is provided as an optional callback that gets executed at this point:

```php
function data_provider_test_image_sizes_equality(): array {
	return array(
		'no_filter' => array(
			'add_filter' => null,
		),
		'with_filter' => array(
			'add_filter' => static function () {
				add_filter(
					'wp_content_img_tag',
					static function (
						string $content
					): string {
						$processor = new WP_HTML_Tag_Processor( $content );
						if ( ! $processor->next_tag( array( 'tag_name' => 'img' ) ) ) {
							return $content;
						}
						$processor->set_attribute( 'sizes', '(max-width: 333px) 100vw, 333px' );
						return $processor->get_updated_html();
					}
				);
			},
		)
	);
}
```

So in the test you can just do:

```php
if ( $add_filter instanceof Closure ) {
	$add_filter();
}
```

westonruter reviewed (commented):
Looking good! Almost there.

mukeshpanchal27 committed:
Merge branch 'trunk' into fix/1349-sizes-picture-element

mukeshpanchal27 committed:
Add assert check for IMG

mukeshpanchal27 committed:
Use dataprovide for unit test

mukeshpanchal27 committed:
More update on tests

mukeshpanchal27 commented:
Thanks @westonruter @adamsilverstein (Joe OoO) for the feedback. PR ready for review.

westonruter commented on plugins/webp-uploads/tests/test-picture-element.php:
```diff
@@ -125,4 +140,121 @@ public function data_provider_it_should_maybe_wrap_images_in_picture_element():
+			'with_filter' => array(
+				'add_filter' => static function (): void {
+					add_filter(
+						'wp_content_img_tag',
+						static function (
+							string $content
+						): string {
+							$processor = new WP_HTML_Tag_Processor( $content );
+							if ( ! $processor->next_tag( array( 'tag_name' => 'img' ) ) ) {
+								return $content;
+							}
```
If there is no `IMG` in the `$content`, this would be bad, right? I don't think this should return silently.

```suggestion
						'wp_content_img_tag',
						function ( string $content ): string {
							$processor = new WP_HTML_Tag_Processor( $content );
							$this->assertTrue( $processor->next_tag( array( 'tag_name' => 'img' ) ) );
```

mukeshpanchal27 replied:
It generate PHPStan error: https://github.com/WordPress/performance/pull/1354/commits/61a488f10c0b9408bdc25820a2abf2f01218ebfd

westonruter commented on plugins/webp-uploads/tests/test-picture-element.php:
```diff
@@ -125,4 +140,121 @@ public function data_provider_it_should_maybe_wrap_images_in_picture_element():
+	/**
+	 * Data provider for it_should_maybe_wrap_images_in_picture_element.
+	 *
+	 * @return array<string, array<string, (Closure)|null>>
```
```suggestion
	 * @return array<string, array{ add_filter: Closure|null }>
```

westonruter reviewed (approved):
Just a couple nits, but pre-approving as LGTM.

adamsilverstein reviewed (approved):
Excellent!

mukeshpanchal27 committed:
Address unit test feedback

Co-authored-by: Weston Ruter <westonruter@google.com>

mukeshpanchal27 committed:
Fix phpstan error

mukeshpanchal27 committed:
Correct docblock

Co-authored-by: Weston Ruter <westonruter@google.com>

mukeshpanchal27 committed:
Merge branch 'trunk' into fix/1349-sizes-picture-element
